proto/redpanda/api/console/v1alpha1/security.proto

```diff
@@ -0,0 +1,751 @@
+    (buf.validate.field).string.max_len = 128,
+    (buf.validate.field).string.pattern = "^[a-zA-Z0-9-_]+$",
+    (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+      description: "The role name. i.e. the principal is `RedpandaRole:<role>`",
+      example: "\"admin\""
+    }
+  ];
+
+  // The new role name.
+  string new_name = 2 [
+    (google.api.field_behavior) = REQUIRED,
+    (buf.validate.field).required = true,
+    (buf.validate.field).string.min_len = 1,
+    (buf.validate.field).string.max_len = 128,
+    (buf.validate.field).string.pattern = "^[a-zA-Z0-9-_]+$"
+  ];
+}
+
+// UpdateRoleResponse is the response for UpdateRole.
+message UpdateRoleResponse {
+  // The role name.
+  string role = 1;
+}
+
+// DeleteRoleRequest is the request for DeleteRole.
+message DeleteRoleRequest {
+  // The role name.
+  string role = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (buf.validate.field).required = true,
+    (buf.validate.field).string.min_len = 1,
+    (buf.validate.field).string.max_len = 128,
+    (buf.validate.field).string.pattern = "^[a-zA-Z0-9-_]+$"
+  ];
+
+  // Whether to delete the ACLs bound to the role.
+  optional bool delete_acls = 2;
+}
+
+// DeleteRoleResponse is the response for DeleteRole.
+message DeleteRoleResponse {}
+
+// List role members for a role. That is user principals assigned to that role.
+message ListRoleMembersRequest {
+  // Filter options.
+  message Filter {
+    // Filter results to only mamber named with the prefix.
+    string name_prefix = 1 [(buf.validate.field).string.max_len = 128];
+  }
+
+  // The role name.
+  string role = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (buf.validate.field).required = true,
+    (buf.validate.field).string.min_len = 1,
+    (buf.validate.field).string.max_len = 128,
+    (buf.validate.field).string.pattern = "^[a-zA-Z0-9-_]+$"
+  ];
+
+  // Optional filter.
+  optional Filter filter = 2;
+
+  // Page size.
+  int32 page_size = 3 [
+    (buf.validate.field).int32 = {
+      gte: -1,
+      lte: 1000
+    },
+    (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+      description: "Limit the paginated response to a number of items. Defaults to 100. Use -1 to disable pagination.",
+      minimum: -1,
+      maximum: 1000
+    }
+  ];
+
+  // Value of the next_page_token field returned by the previous response.
+  // If not provided, the system assumes the first page is requested.
+  string page_token = 4;
+}
+
+// ListRoleMembersResponse is the response for ListRoleMembers.
+message ListRoleMembersResponse {
+  // Members assigned to the role.
+  repeated RoleMembership members = 1;
+}
+
+// RoleMembership is the role membership.
+message RoleMembership {
+  // The name of the principal assigned to the role.
+  string name = 1;
+
+  // The principal type.
+  Security.PrincipalType principal_type = 2;
+}
+
+// UpdateRoleMembershipRequest is the request to UpdateRoleMembership.
+message UpdateRoleMembershipRequest {
```

Comment:
Just a thought, but the semantics for updating resources here is very different to what we do on the Kafka API for topic/broker configs for example. I assume you sticked closely to how the Redpanda API works. Personally I'd prefer consistency across endpoints for such things, but I'm unsure how well it would work out across all the upstream APIs we have to support.

Reply:
@bojand

Reply:
Okay let's remain close to what we get from upstream and we'll revisit this (having consistent semantics across all endpoints) at some point in the future, but let's always keep it in mind when adding new API endpoints.